cni/cmd/install/main.go

```diff
@@ -0,0 +1,316 @@
+	file, found := lookForValidConfig(files)
+	if found {
+		return file
+	}
+
+	files, _ = filepath.Glob(mountedCNINetDir + "/*.conflist")
+	file, found = lookForValidConfig(files)
+	if found {
+		return file
+	}
+
+	// probably should return an error
+	return ""
+}
+
+func lookForValidConfig(files []string) (string, bool) {
+	for _, file := range files {
+		found := isFileAValidConfig(file)
+		if found {
+			return file, true
+		}
+	}
+	return "", false
+}
+
+func isFileAValidConfig(file string) bool {
+	var parsed map[string]interface{}
+	contents, _ := ioutil.ReadFile(file)
+	// this is probably going to be rewritten to not use `jq` at all
+	json.Unmarshal(contents, &parsed)
+	query, _ := gojq.Parse(`has("type")`)
+	iterator := query.Run(parsed)
+	v, ok := iterator.Next()
+	if !ok {
+		return false
+	}
+	logger.Default.Info("checking file", zap.String("file", file))
+	if v.(bool) == true {
+		return true
+	}
+	return false
+}
+
+func check_install(mountedCNINetDir string) {
+	// todo: implement
+}
+
+func cleanup() {
+	// todo implement
+}
+
+func install() error {
+	hostCniNetDir := env.GetString("CNI_NET_DIR", "/etc/cni/net.d")
+	kubecfgName := env.GetString("KUBECFG_FILE_NAME", "ZZZ-kuma-cni-kubeconfig")
+	cfgCheckInterval, _ := env.GetInt("CFGCHECK_INTERVAL", 1)
+	chainedCniPlugin, _ := env.GetBool("CHAINED_CNI_PLUGIN", true)
+	mountedCniNetDir := env.GetString("MOUNTED_CNI_NET_DIR", "/host/etc/cni/net.d")
+	cniConfName := env.GetString("CNI_CONF_NAME", find_cni_conf_file(mountedCniNetDir))
+	serviceAccountPath := "/var/run/secrets/kubernetes.io/serviceaccount"
+
+	copyBinaries(cniConfName)
+	err := prepareKubeconfig(mountedCniNetDir, kubecfgName, serviceAccountPath)
+	if err != nil {
+		return err
+	}
+	err = prepareKumaCniConfig(mountedCniNetDir, hostCniNetDir, kubecfgName, serviceAccountPath, cniConfName, chainedCniPlugin)
+	if err != nil {
+		return err
+	}
+
+	shouldSleep, _ := env.GetBool("SLEEP", true)
+
+	for shouldSleep == true {
+		time.Sleep(time.Duration(cfgCheckInterval) * time.Second)
+		check_install(mountedCniNetDir)
+	}
+
+	return nil
+}
+
+func prepareKumaCniConfig(mountedCniNetDir, hostCniNetDir, kubecfgName, serviceAccountPath, cniConfName string, chainedCniPlugin bool) error {
+	rawConfig := env.GetString("CNI_NETWORK_CONFIG", "")
+	kubeconfigFilePath := hostCniNetDir + "/" + kubecfgName
+
+	config := strings.Replace(rawConfig, "__KUBECONFIG_FILEPATH__", kubeconfigFilePath, 1)
+	logger.Default.Debug("config after replace", zap.String("config", config))
+
+	serviceAccountToken, err := ioutil.ReadFile(serviceAccountPath + "/token")
+	if err != nil {
+		return err
+	}
+	config = strings.Replace(config, "__SERVICEACCOUNT_TOKEN__", string(serviceAccountToken), 1)
+
+	if chainedCniPlugin {
+		err := setupChainedPlugin(mountedCniNetDir, cniConfName, config)
+		if err != nil {
+			logger.Default.Error("unable to setup kuma cni as chained plugin", zap.Error(err))
+			return err
```

Comment:
https://github.com/kumahq/kuma/pull/4445/files#diff-03543e676073151a7aba162f34cbd764dc512a896e8f7a5bf2880bea0fe0bb85R56

also in other places